subprojects/core/src/main/java/org/gradle/internal/progress/DefaultBuildOperationExecutor.java

```diff
@@ -16,157 +16,243 @@
 
 package org.gradle.internal.progress;
 
+import com.google.common.collect.Lists;
+import org.apache.commons.lang.StringUtils;
 import org.gradle.api.Action;
+import org.gradle.api.GradleException;
 import org.gradle.api.Nullable;
 import org.gradle.api.Transformer;
-import org.gradle.internal.Transformers;
+import org.gradle.internal.Cast;
+import org.gradle.internal.SystemProperties;
 import org.gradle.internal.UncheckedException;
+import org.gradle.internal.concurrent.ExecutorFactory;
 import org.gradle.internal.concurrent.GradleThread;
+import org.gradle.internal.concurrent.StoppableExecutor;
+import org.gradle.internal.exceptions.DefaultMultiCauseException;
 import org.gradle.internal.logging.events.OperationIdentifier;
 import org.gradle.internal.logging.progress.ProgressLogger;
 import org.gradle.internal.logging.progress.ProgressLoggerFactory;
+import org.gradle.internal.operations.BuildOperation;
 import org.gradle.internal.operations.BuildOperationContext;
+import org.gradle.internal.operations.BuildOperationExecutor;
+import org.gradle.internal.operations.BuildOperationQueue;
+import org.gradle.internal.operations.BuildOperationQueueFactory;
+import org.gradle.internal.operations.BuildOperationQueueFailure;
+import org.gradle.internal.operations.BuildOperationWorker;
+import org.gradle.internal.operations.CallableBuildOperation;
+import org.gradle.internal.operations.MultipleBuildOperationFailures;
+import org.gradle.internal.operations.RunnableBuildOperation;
 import org.gradle.internal.time.TimeProvider;
+import org.gradle.util.CollectionUtils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
 import java.util.concurrent.atomic.AtomicLong;
 
+//TODO move to base-services once the ProgressLogger dependency is removed
 public class DefaultBuildOperationExecutor implements BuildOperationExecutor {
-
     private static final Logger LOGGER = LoggerFactory.getLogger(DefaultBuildOperationExecutor.class);
+    private static final String LINE_SEPARATOR = SystemProperties.getInstance().getLineSeparator();
 
     private final BuildOperationListener listener;
     private final TimeProvider timeProvider;
     private final ProgressLoggerFactory progressLoggerFactory;
+    private final BuildOperationQueueFactory buildOperationQueueFactory;
+    private final StoppableExecutor fixedSizePool;
+
     private final AtomicLong nextId = new AtomicLong();
-    private final ThreadLocal<OperationDetails> currentOperation = new ThreadLocal<OperationDetails>();
+    private final ThreadLocal<BuildOperationState> currentOperation = new ThreadLocal<BuildOperationState>();
+    private final Set<Object> runningOperationIds = Collections.synchronizedSet(new HashSet<Object>());
```

Comment:
❌ We shouldn't introduce locking into this hot code path. The previous solution with `state.running` only used a volatile read, which is more efficient. We should either go back to that or use a concurrent map.